internal/validitywindow/client.go

```diff
@@ -0,0 +1,181 @@
+	resultChan := make(chan FetchResult[B], 100)
+
+	// Start fetching in a separate goroutine
+	go func() {
+		c.checkpointLock.Lock()
+		c.checkpoint = checkpoint{
+			parentID:   id,
+			nextHeight: height,
+			timestamp:  timestamp,
+		}
+		c.checkpointLock.Unlock()
+		req := &BlockFetchRequest{MinTimestamp: minTimestamp.Load()}
+
+		for {
+			reqCtx, cancel := context.WithTimeout(ctx, requestTimeout)
+
+			select {
+			case <-ctx.Done():
+				resultChan <- FetchResult[B]{Err: ctx.Err()}
+				cancel()
+				return
+			default:
+			}
+
+			c.checkpointLock.RLock()
+			lastCheckpoint := c.checkpoint
+			c.checkpointLock.RUnlock()
+
+			if lastCheckpoint.timestamp <= minTimestamp.Load() {
+				close(resultChan)
+				cancel()
+				return
+			}
+
+			nodeID := c.sampleNodeID(ctx)
+			if nodeID.Compare(ids.EmptyNodeID) == 0 {
+				continue
+			}
+
+			req.BlockHeight = lastCheckpoint.nextHeight
+			err := c.client.AppRequest(reqCtx, nodeID, req, func(ctx context.Context, nodeID ids.NodeID, response *BlockFetchResponse, err error) {
+				// Handle response
+				if err != nil {
+					fmt.Printf("Handler received error: %v\n", err)
+					resultChan <- FetchResult[B]{Err: err}
+					return
+				}
+
+				respBlocks := response.Blocks
+				if len(respBlocks) == 0 {
+					resultChan <- FetchResult[B]{Err: fmt.Errorf("node=%s: %w", nodeID, errEmptyResponse)}
+					return
+				}
+
+				c.checkpointLock.RLock()
+				expectedParentID := c.checkpoint.parentID
+				c.checkpointLock.RUnlock()
+
+				for _, raw := range respBlocks {
+					block, parseErr := c.parser.ParseBlock(ctx, raw)
+					if parseErr != nil {
+						resultChan <- FetchResult[B]{Err: errInvalidBlock}
+						return
+					}
+
+					if expectedParentID != block.GetID() {
+						resultChan <- FetchResult[B]{Err: fmt.Errorf("expectedParentID=%s got=%s: %w", expectedParentID, block.GetID(), errInvalidBlock)}
+						return
+					}
+					expectedParentID = block.GetParent()
+
+					select {
+					// try to write
+					case resultChan <- FetchResult[B]{Block: maybe.Some(block)}:
+						// Update checkpoint
+						c.checkpointLock.Lock()
+						c.checkpoint.parentID = block.GetParent()
+						c.checkpoint.timestamp = block.GetTimestamp()
+						c.checkpoint.nextHeight = block.GetHeight() - 1
+						c.checkpointLock.Unlock()
+					case <-ctx.Done():
+						resultChan <- FetchResult[B]{Err: ctx.Err()}
+						return
+					default:
+						resultChan <- FetchResult[B]{Err: errChannelFull}
+						return
+					}
+				}
+			})
+			if err != nil {
+				cancel()
+				resultChan <- FetchResult[B]{Err: fmt.Errorf("fetch error from node=%s: %w", nodeID, err)}
+			}
+			time.Sleep(500 * time.Millisecond)
+		}
+	}()
+
+	return resultChan
+}
+
+func (c *BlockFetcherClient[B]) sampleNodeID(ctx context.Context) ids.NodeID {
+	nodes := c.sampler.Sample(ctx, numSampleNodes)
+	for _, nodeID := range nodes {
+		if nodeID.Compare(ids.EmptyNodeID) != 0 {
+			return nodeID
+		}
+	}
```

Comment:
Do we need this check? It should only contain valid nodeIDs, so I'd think the only thing we may need to check is that the length is non-zero